File: services/tenantd/csharp/src/CtlFlow.Tenancy.Tenantd.Service/Aggregation/Documents/TenantDocument.cs
namespace CtlFlow.Tenancy.Tenantd.Service.Aggregation.Documents;

internal sealed class TenantDocument
{
    public required string ApiVersion { get; init; }

    public required string Kind { get; init; }

    public required ObjectMetaDocument Metadata { get; init; }

    public required TenantSpecDocument Spec { get; init; }

    public ResourceStatusDocument? Status { get; init; }
}

internal sealed class TenantSpecDocument
{
    public required string DisplayName { get; init; }

    public required ExternalTenantAddressDocument Address { get; init; }

    public required InitialAdministratorDocument InitialAdministrator
    {
        get;
        init;
    }

    public required BaselinePackageDocument[] BaselinePackages { get; init; }
}

internal sealed class TenantListDocument
{
    public required string ApiVersion { get; init; }

    public required string Kind { get; init; }

    public required ListMetaDocument Metadata { get; init; }

    public required TenantDocument[] Items { get; init; }
}

File: services/tenantd/tests/support/tenant-document.ts
export interface TenantDocument {
  readonly apiVersion: "tenancy.ctlflow.com/v1alpha1";
  readonly kind: "Tenant";
  readonly metadata: {
    readonly name: string;
    readonly resourceVersion: string;
    readonly creationTimestamp: string;
  };
  readonly spec: {
    readonly displayName: string;
    readonly address: {
      readonly authority: string;
      readonly pathPrefix: string;
    };
    readonly initialAdministrator: {
      readonly displayName: string;
      readonly loginIdentifier: string;
      readonly identityLink?: {
        readonly providerId: string;
        readonly providerSubject: string;
      };
    };
    readonly baselinePackages: readonly {
      readonly packageId: string;
      readonly packageVersion: string;
    }[];
  };
  readonly status: {
    readonly lifecycle: string;
    readonly revision: number;
    readonly provisioningGeneration: number;
    readonly currentOperation?: {
      readonly id: string;
      readonly kind: string;
    };
    readonly conditions: readonly {
      readonly owner: string;
      readonly state: string;
      readonly ownerRevision?: number;
      readonly reason?: string;
      readonly lastTransitionTime: string;
    }[];
  };
}
